File: algorithm/leetcodeObjectiveC/3sum.m
// 15. 3Sum
// Difficulty: Medium
// Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0?
// Find all unique triplets in the array which gives the sum of zero.

// Note: The solution set must not contain duplicate triplets.

// For example, given array S = [-1, 0, 1, 2, -1, -4],

// A solution set is:
// [
//   [-1, 0, 1],
//   [-1, -1, 2]
// ]

// Time:  O(n^2)
// Space: O(1)

#import <Foundation/Foundation.h>

NSArray* threeSum(NSArray* nums) {
  NSMutableArray* result = @[].mutableCopy;
  NSArray* sorted = [nums sortedArrayUsingComparator:
                          ^NSComparisonResult(id obj1, id obj2){
                            return [obj1 compare:obj2];
                          }];
  NSInteger target = 0;
  
  for (NSInteger i = 0; i < [sorted count] - 2; ++i) {
    if (i == 0 || [sorted[i] integerValue] != [sorted[i - 1] integerValue]) {
      for (NSInteger j = i + 1, k = [sorted count] - 1; j < k; ) {
        if (j - 1 > i && sorted[j] == sorted[j - 1]) {
          ++j;
        } else if (k + 1 < [sorted count] && [sorted[k] integerValue]== [sorted[k + 1] integerValue]) {
          --k;
        } else {
          NSInteger sum = [sorted[i] integerValue] + [sorted[j] integerValue] + [sorted[k] integerValue];
          if (sum > target) {
            --k;
          } else if (sum < target) {
            ++j;
          } else {
            [result addObject:@[sorted[i], sorted[j], sorted[k]]];
            ++j;
            --k;
          }
        }
      }
    }
  }
  
  return result;
}

File: zZ/javascript/leetcode/3-sum.js
// 15. 3Sum

// Given an array nums of n integers, are there elements a, b, c in nums such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

// Note:

// The solution set must not contain duplicate triplets.

// Example:

// Given array nums = [-1, 0, 1, 2, -1, -4],

// A solution set is:
// [
//   [-1, 0, 1],
//   [-1, -1, 2]
// ]

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    if (!isArray(nums)) {
        throw "invalid input array"
    }
    
    nums.sort((a, b) => a - b)
    
    const result = []
    for (let i = 0; i < nums.length - 2; ++i) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue
        }
        
        for (let j = i + 1, k = nums.length - 1; j < k;) {
            if (j > i + 1 && nums[j] === nums[j - 1]) {
                ++j
                continue
            }
            
            if (k < nums.length - 1 && nums[k] === nums[k + 1]) {
                --k
                continue
            }
            
            const sum = nums[i] + nums[j] + nums[k]
            
            if (sum === 0) {
                result.push([nums[i], nums[j], nums[k]])
                ++j
                --k
            } else if (sum > 0) {
                --k
            } else {
                ++j
            }
        }
    }
    
    return result
};

function isArray(item) {
    return Array.isArray(item)
}